docs/styleguides/python.md

```diff
@@ -0,0 +1,173 @@
+# DCOS Python Style Guide
+###### Version 0.1
+###### April-6-2017
+
+For Python source code within DCOS, we prefer to keep Google Python Style Guide Revision 2.59 as the base and **extend** it with
+some changes that will make DCOS codebase look modern, consistent and easy to follow.
+
+##### Rationale
+
+With any style guide, the primary requirement is to adopt one and follow it consistently. With our choices below, we can use tools available in the ecosystem to our advantage, enforce consistency in the codebase, creating and publishing automatically generated documentation for our code.
+
+##### Versioned document
+
+Styleguide _conventions_ are not set in stone. We can improve upon this or remove inconvenient styles by creating a PR, discussing the merits or demerits to arrive at a consensus.
+
+#### Base Style Guide
+
+* [Google Python Style Guide Revision 2.59](https://google.github.io/styleguide/pyguide.html)
+
+#### Summary of Style Requirements DCOS which extend Base Style Guide and are specific to DCOS.
+
+* Import Requirements
+    * Adopt the base style guide preferences.
+
+        ```
+        Use import x for importing packages and modules.
+        Use from x import y where x is the package prefix and y is the module name with no prefix.
+        Use from x import y as z if two modules named y are to be imported or if y is an inconveniently long name.
+        ```
+    * Use `__init__.py` only to represent python packages.
+        * Avoid importing of objects in `__init__.py` for readability.
+* Import formatting
+
+    * The types of imports should be separated by single line and order is should be.
+        * standard library imports.
+        * third-party imports.
+        * application specific imports.
+
+    * The `from` style import and `import` statements should be separated by single line too.
+    * Each group of these imports should be sorted.
+
+For Example
+
+```
+import json
+import uuid
+
+from subprocess import CalledProcessError
+from typing import Callable, List
+
+import retrying
+
+from retrying import retry, RetryError
+from teamcity.messages import TeamcityServiceMessages
+
+import test_util.aws
+import test_util.cluster
+
+from pkgpanda.util import logger, write_string
+```
+
+* [Packages](https://google.github.io/styleguide/pyguide.html?showone=Packages#Packages)
+    * Import each module using the full pathname location of the module.
+    * No Relative Imports.
+
+* Maximum line length is 120 characters.
```

Comment:
> Our existing code base is 120 col width.

Some of the codebase is 80 col (AR test harness code, Bouncer, dcos-security-cli, dcos-ca-pylib to name a few), some of the code is 120 lines (most notably genconf).

>  as I feel 80, in this day and age is far too small give how much lateral screen space is available.

Yes, as long as you working at your desk. As soon as you start travelling and/or working away from your desk, you are stuck with a small laptop screen. One can try to decrease the font size, but this is burdensome for one's eyes.

Also, looking at some other cody style guidelines:
* from pep8(https://www.python.org/dev/peps/pep-0008/#maximum-line-length):
```
Limit all lines to a maximum of 79 characters.
```
* from linux kernel style guideline (https://01.org/linuxgraphics/gfx-docs/drm/process/coding-style.html) (it's my personal favourite):
```
The limit on the length of lines is 80 columns and this is a strongly preferred limit.
```
* Drupal (PHP codebase) coding style guideline (https://www.drupal.org/docs/develop/standards/coding-standards#linelength):
```
In general, all lines of code should not be longer than 80 characters.
```
* and finally Django project has a nice compromise I think:
```
As an exception to PEP 8 is our rules on line lengths. Don’t limit lines of code to 79 characters if it means the code looks significantly uglier or is harder to read. We allow up to 119 characters as this is the width of GitHub code review; anything longer requires horizontal scrolling which makes review more difficult. This check is included when you run flake8. Documentation, comments, and docstrings should be wrapped at 79 characters, even though PEP 8 suggests 72.
```

I was wondering if we could rephrase our rules to follow the ones from Django project. For me personally, it would be great if each project could choose it's own max line length as long as it's smaller than the global limit of 120.